src/anemoi/training/losses/weightedloss.py

```diff
@@ -0,0 +1,155 @@
+# (C) Copyright 2024 ECMWF.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+#
+
+from __future__ import annotations
+
+import logging
+from abc import ABC
+from abc import abstractmethod
+
+import torch
+from torch import nn
+
+LOGGER = logging.getLogger(__name__)
+
+
+class BaseWeightedLoss(nn.Module, ABC):
+    """Node-weighted general loss."""
+
+    def __init__(
+        self,
+        node_weights: torch.Tensor,
+        loss_scaling: torch.Tensor | None = None,
+        ignore_nans: bool = False,
+    ) -> None:
+        """Node- and feature_weighted Loss.
+
+        Exposes:
+        - self.avg_function: torch.nanmean or torch.mean
+        - self.sum_function: torch.nansum or torch.sum
+        depending on the value of `ignore_nans`
+
+        Registers:
+        - self.node_weights: torch.Tensor of shape (N, )
+        - self.loss_scaling: torch.Tensor if given
+
+        Parameters
+        ----------
+        node_weights : torch.Tensor of shape (N, )
+            Weight of each node in the loss function
+        loss_scaling : Optional[torch.Tensor], optional
+            precomputed, per-variable weights, by default None
+        ignore_nans : bool, optional
+            Allow nans in the loss and apply methods ignoring nans for measuring the loss, by default False
+
+        """
+        super().__init__()
+
+        self.avg_function = torch.nanmean if ignore_nans else torch.mean
+        self.sum_function = torch.nansum if ignore_nans else torch.sum
+
+        self.register_buffer("node_weights", node_weights, persistent=True)
+        self.register_buffer("loss_scaling", loss_scaling, persistent=True)
+
+    def scale_by_loss_scaling(
+        self,
+        x: torch.Tensor,
+        feature_indices: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        """Scale a tensor by the loss_scaling.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Tensor to be scaled, shape (bs, (optional_ensemble), lat*lon, n_outputs)
```

Comment:
We shouldn't have an optional ensemble dimension here. It says optional but actually the squash doesn't work if you don't have an ensemble dimension

Reply:
I believe it is always there, but just sometimes 1

Reply:
OK I think this is maybe not ideal but happy to address this in a subsequent PR